File: contracts/session.py
"""Session record schema written by C11 LocalSessionLog at scenario debrief (FR9).

Per design §6f Backend → File system contract row.
Mirror of contracts/session.ts.
"""
from datetime import datetime
from typing import Literal, Optional
from uuid import UUID
from pydantic import BaseModel, Field


class CompressionEvent(BaseModel):
    ts: int
    depth: float
    rate: int
    is_adequate: bool
    recoil_complete: bool


class DecisionEvent(BaseModel):
    decision_point: str
    choice_made: Optional[str]
    correct_choice: str
    is_correct: bool
    time_to_decide_seconds: float


class ErrorEvent(BaseModel):
    ts: int
    component: str
    error_code: str
    fallback_used: Optional[str] = None


class SessionRecord(BaseModel):
    """Full per session record. Written once at scenario debrief to data/local/<session_id>.json."""

    session_id: UUID
    scenario_type: str
    started_at: datetime
    ended_at: datetime
    patient_survived: bool
    overall_score: float
    response_time_seconds: float
    compressions: list[CompressionEvent] = Field(default_factory=list)
    decisions: list[DecisionEvent] = Field(default_factory=list)
    errors: list[ErrorEvent] = Field(default_factory=list)

File: contracts/session.ts
// Session record schema written by C11 LocalSessionLog at scenario debrief (FR9).
// Per design §6f Backend → File system contract row.
// Mirror of contracts/session.py.
import { z } from 'zod';

export const CompressionEventSchema = z.object({
  ts: z.number().int(),
  depth: z.number(),
  rate: z.number().int(),
  is_adequate: z.boolean(),
  recoil_complete: z.boolean(),
});

export const DecisionEventSchema = z.object({
  decision_point: z.string(),
  choice_made: z.string().nullable(),
  correct_choice: z.string(),
  is_correct: z.boolean(),
  time_to_decide_seconds: z.number(),
});

export const ErrorEventSchema = z.object({
  ts: z.number().int(),
  component: z.string(),
  error_code: z.string(),
  fallback_used: z.string().nullable(),
});

export const SessionRecordSchema = z.object({
  session_id: z.string().uuid(),
  scenario_type: z.string(),
  started_at: z.string().datetime(),
  ended_at: z.string().datetime(),
  patient_survived: z.boolean(),
  overall_score: z.number(),
  response_time_seconds: z.number(),
  compressions: z.array(CompressionEventSchema),
  decisions: z.array(DecisionEventSchema),
  errors: z.array(ErrorEventSchema),
});

export type CompressionEvent = z.infer<typeof CompressionEventSchema>;
export type DecisionEvent = z.infer<typeof DecisionEventSchema>;
export type ErrorEvent = z.infer<typeof ErrorEventSchema>;
export type SessionRecord = z.infer<typeof SessionRecordSchema>;
